src/stores/external-source.ts

```diff
@@ -0,0 +1,185 @@
+);
+export const planDerivationGroupLinks = gqlSubscribable<PlanDerivationGroup[]>(
+  gql.SUB_PLAN_DERIVATION_GROUP,
+  {},
+  [],
+  null,
+);
+
+// this tracks each user's view of the sources. if something exists in externalSources that isn't in usersSeenSources, it's treated as unseen, and vice versa for deleted.
+export const usersSeenSourcesRaw = gqlSubscribable<{
+  derivation_group: string;
+  external_source_name: string;
+  external_source_type: string;
+  id: number;
+  username: string;
+}[]>(
+  gql.SUB_SEEN_SOURCES,
+  {},
+  [],
+  null
+);
+
+/* Derived. */
+export const selectedPlanDerivationGroupNames = derived(
+  [planDerivationGroupLinks, planId],
+  ([$planDerivationGroupLinks, $planId]) =>
+    $planDerivationGroupLinks.filter(link => link.plan_id === $planId).map(link => link.derivation_group_name),
+);
+export const selectedPlanDerivationGroupEventTypes = derived(
+  [derivationGroups, selectedPlanDerivationGroupNames],
+  ([$derivationGroups, $selectedPlanDerivationGroupIds]) => {
+    return $derivationGroups
+      .filter(derivationGroup => $selectedPlanDerivationGroupIds.includes(derivationGroup.name))
+      .map(derivationGroup => derivationGroup.event_types)
+      .reduce((acc, curr) => acc.concat(curr), []);
+  },
+);
+export const usersSeenSources = derived(
+  [usersSeenSourcesRaw, externalSources],
+  ([$usersSeenSourcesRaw, $externalSources]) => {
+    const res: Record<string, UserSeenEntryWithDate[]> = {};
+    for (const entry of $usersSeenSourcesRaw) {
+      const change_date = $externalSources.find(externalSource => externalSource.pkey.derivation_group_name === entry.derivation_group && externalSource.pkey.key === entry.external_source_name)?.created_at ?? ""
+      if (res[entry.username]) {
+        res[entry.username].push({
+          change_date,
+          derivation_group_name: entry.derivation_group,
+          key: entry.external_source_name,
+          source_type_name: entry.external_source_type,
+        });
+      } else {
+        res[entry.username] = [
+          {
+            change_date,
+            derivation_group_name: entry.derivation_group,
+            key: entry.external_source_name,
+            source_type_name: entry.external_source_type,
+          },
+        ];
+      }
+    }
+    return res;
+  }
+);
+
+/* Helper Functions. */
+export function resetExternalSourceStores() {
+  createExternalSourceError.set(null);
+  createExternalSourceTypeError.set(null);
+  createDerivationGroupError.set(null);
+  derivationGroupPlanLinkError.set(null);
+}
+
+function transformExternalSources(
+  externalSources: {
+    created_at: string;
+    derivation_group_name: string;
+    end_time: string;
+    key: string;
+    source_type_name: string;
+    start_time: string;
+    valid_at: string;
+  }[],
+): ExternalSourceSlim[] {
+  const completeExternalSourceSlim: ExternalSourceSlim[] = [];
+  if (externalSources !== null && externalSources !== undefined) {
```

Comment:
is this check necessary if the argument type for `externalSources` isn't specified as being optional?

Reply:
Resolved.